### PR TITLE
Refactor training build options

### PR DESCRIPTION
**Pull request by askhade**
### Description
1. Renames all references of on device training to training apis. This is to keep the naming general. Nothing really prevents us from using the same apis on servers\non-edge devices.
2. Update ENABLE_TRAINING option: With this PR when this option is enabled, training apis and torch interop is also enabled. 
3. Refactoring for onnxruntime_ENABLE_TRAINING_TORCH_INTEROP option: 
   -  Removed user facing option
   - Setting onnxruntime_ENABLE_TRAINING_TORCH_INTEROP to ON when onnxruntime_ENABLE_TRAINING is ON as we always build with torch interop. 

Once this PR is merged when --enable_training is selected we will do a "FULL Build" for training (with all the training entry points and features). 
Training entry points include:
1. ORTModule
2. Training APIs

Features include:
1. ATen Fallback
2. All Training OPs includes communication and collectives
3. Strided Tensor Support
4. Python Op (torch interop)
5. ONNXBlock (Front end tools for training artifacts prep when using trianing apis)

### Motivation and Context
Intention is to simply the options for building training enabled builds. This is part of the larger work item to create dedicated build for learning on the edge scenarios with just training apis enabled.

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+trigger: none
```
Comment (baijumeswani) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml:1]: Was the addition of this yml intentional or was it meant to be a replacement for tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml ?

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml [hunk 2]
```diff
@@ -15,30 +15,30 @@ jobs:
     parameters:
       RunDockerBuildArgs: |
         -o ubuntu20.04 -d gpu -e \
-        -t onnxruntime_ondevicetraining_tests_image \
+        -t onnxruntime_training_apis_tests_image \
         -x " \
           --config RelWithDebInfo \
           --enable_training \
-          --enable_training_on_device \
+          --enable_training_apis \
```
Comment (baijumeswani) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml:22]: Considering that enable_training will enable enable_training_apis as well, shall we remove this flag from all the ymls that are currently using this?

Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml:22]: Also, considering enable_training will also enable enable_trianing_apis, shall we get rid of two CI pipelines and build only once?

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 3]
```diff
@@ -33,7 +33,7 @@ public CheckpointState(string checkpointPath)
             }
             else
             {
-                throw new InvalidOperationException("Training is disabled in the current build. Please build ONNXRuntime from source with the build flags enable_training and enable_training_on_device. \n");
+                throw new InvalidOperationException("Training is disabled in the current build. Please build ONNXRuntime from source with the build flags enable_training and enable_training. \n");
```
Comment (Craigacp) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:36]: This says "enable_training and enable_training".

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml (hunk 1 above)
Comment (askhade) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml:1]: I intentionally added a duplicate file. Once this PR is checked in I will update the devops pipeline to use this yml and then delete the original one... If there is a better way to handle this let me know :)

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml (hunk 2 above)
Comment (askhade) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml:22]: I intentionally kept enable_training_apis in this PR... In the next PR where I enable creating a dedicated build for training apis , I am deleting --enable_training flag from this CI yml.

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs (hunk 3 above)
Comment (askhade) [reply csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:36]: will fix

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml (hunk 1 above)
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml:1]: there is a way to do it which requires manually running the pipeline for your PR (by updating the yml file being used for the build). As soon as your PR pipeline is triggered, you can change the pipeline back to using the one in main.
Once the PR is merged, you could change the yml file again.

But up to you, whatever is more convenient.

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml (hunk 2 above)
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml:22]: Do we need a dedicated build for training apis? Shall we just reuse the training CI pipeline which now also builds the training apis?

Comment (askhade) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml:22]: we need to test we are able to build with just enable_training_apis and need to run the relevant tests to validate the builds. Office will be building with just enable_training_apis flag.

Comment (askhade) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-on-device-training.yml:22]: enable_training_apis will be the way to build for learning on the edge scenarios... Office will be first to use this.

File: cmake/CMakeLists.txt [hunk 4]
```diff
@@ -141,11 +141,10 @@ option(onnxruntime_FUZZ_TEST "Enable Fuzz testing" OFF)
 # training options
 option(onnxruntime_ENABLE_NVTX_PROFILE "Enable NVTX profile." OFF)
 option(onnxruntime_ENABLE_MEMORY_PROFILE "Enable memory profile." OFF)
-option(onnxruntime_ENABLE_TRAINING "Enable training functionality." OFF)
-# For on-device training ort build, onnxruntime_ENABLE_TRAINING must be ON.
-option(onnxruntime_ENABLE_TRAINING_ON_DEVICE "Enable training on device." OFF)
+option(onnxruntime_ENABLE_TRAINING "Enable full training functionality. Includes ORTModule and ORT Training APIs" OFF)
+# For building with onnxruntime_ENABLE_TRAINING_APIS, onnxruntime_ENABLE_TRAINING must be ON.
+option(onnxruntime_ENABLE_TRAINING_APIS "Enable ort training apis." OFF)
```
Comment (Craigacp) [inline cmake/CMakeLists.txt:146]: This comment is a bit confusing as later on `onnxruntime_ENABLE_TRAINING` turns on `onnxruntime_ENABLE_TRAINING_APIS`, and the overall idea of this PR seems to involve making `onnxruntime_ENABLE_TRAINING` a higher level flag that enables lots of functionality. Is the support for just enabling `onnxruntime_ENABLE_TRAINING_APIS` coming in a separate PR? I assume the latter is the option I should be looking for in Java as we can't interop with the `ORTModule` bits in pytorch.

Comment (askhade) [reply cmake/CMakeLists.txt:146]: Yes, the support for just enabling onnxruntime_ENABLE_TRAINING_APIS is coming in a separate PR.

File: orttraining/orttraining/test/training_api/core/training_api_tests.cc [hunk 5]
```diff
@@ -150,7 +150,7 @@ void TestLRSchduler(const std::string& test_file_name, float initial_lr, int64_t
 
   /// Load test data for learning rate schedulers.
   auto data_dir = MODEL_FOLDER "test_data_generation/lr_scheduler/";
```
Comment (snnn) [inline orttraining/orttraining/test/training_api/core/training_api_tests.cc:152]:
```
  constexpr const ORTCHAR_T* data_dir = MODEL_FOLDER ORT_TSTR("test_data_generation/lr_scheduler/");
```
Then in the next line you don't need to convert this string from char* to wchar_t*. However, when you define the MODEL_FOLDER macro, you also need to do it in such a way.

Comment (askhade) [reply orttraining/orttraining/test/training_api/core/training_api_tests.cc:152]: Changed this a little... please review again

**Review by snnn — APPROVED** [review]
LGTM.

**Review by baijumeswani — DISMISSED** [review]

**Review by baijumeswani — DISMISSED** [review]

**Review by baijumeswani — APPROVED** [review]

**Review by jchen351 — APPROVED** [review]

**Pull request merged**